File: 2023/python/day02b/main.py
import os

f = open(os.path.join(os.path.dirname(__file__), 'input.txt'), "r")

input_data = f.read().splitlines()
sum = 0
for line in input_data:
    d = {
        "red": 0,
        "green": 0,
        "blue": 0,
    }

    game, rounds = line.split(": ")
    id = int(game.split(' ')[1])

    rounds = rounds.split("; ")
    cont = False
    for round in rounds:
        cubes = round.split(", ")
        for cube in cubes:
            cube = cube.split(' ')
            color = cube[1]
            number = int(cube[0])
            if d[color] < number:
                d[color] = number

    sum+= d["red"] * d["blue"] * d["green"]

print(sum)

f.close()

File: 2023/rust/day02b/src/main.rs
fn main() {
    let input: Vec<_> = include_str!("input.txt")
        .split('\n')
        .collect();
    let mut sum: i32 = 0;

    for line in input {
        let game: Vec<&str> = line.split(": ").collect();
        let records: Vec<&str> = game[1].split("; ").collect();

        let mut min_red: i32 = 0;
        let mut min_green: i32 = 0;
        let mut min_blue: i32 = 0;
        for record in records {
            let cubes: Vec<&str> = record.split(", ").collect();
            for cube in cubes {
                let components: Vec<&str> = cube.split(' ').collect();
                let color: &str = components[1];
                let number: i32 = components[0].parse::<i32>().unwrap();
                match color {
                    "green" => {
                        if number > min_green {
                            min_green = number;
                        }
                    },
                    "red" => {
                        if number > min_red {
                            min_red = number;
                        }
                    },
                    "blue" => {
                        if number > min_blue {
                            min_blue = number;
                        }
                    }
                    _ => panic!("Invalid color")
                }
            }
        }

        let power: i32 = min_green * min_red * min_blue;
        sum += power;
    }
    println!("{}", sum);
}